File: analyzers/tests/SonarAnalyzer.Test/TestCases/CatchRethrow.vb
Imports System

Namespace Tests.TestCases
    Class CatchRethrow
        Private Sub doSomething()
            Throw New NotSupportedException()
        End Sub

        Public Sub Test()
            Dim someWronglyFormatted = 45

            Try
                doSomething()
            Catch exc As Exception 'Noncompliant
                Throw
            End Try

            Try
                doSomething()
            Catch exc As ArgumentException 'Noncompliant
                Throw
            End Try

            Try
                doSomething()
            Catch exc As ArgumentException 'Noncompliant
                Throw
            Catch exc As NotSupportedException 'Noncompliant
                Throw
            End Try

            Try
                doSomething()
            Catch exc As ArgumentException
                Throw
            Catch
                Console.WriteLine("")
                Throw
            End Try

            Try
                doSomething()
            Catch exc As ArgumentException 'Noncompliant
                Throw
            Catch exc As NotSupportedException
                Console.WriteLine("")
                Throw
            End Try

            Try
                doSomething()
            Catch exc As ArgumentException When True
                Throw
            Catch exc As NotSupportedException
                Console.WriteLine("")
                Throw
            End Try

            Try
                doSomething()
            Catch exc As ArgumentException When True
                Throw
            End Try

            Try
                doSomething()
            Catch exc As NotSupportedException 'Noncompliant
                Throw
            Finally
            End Try

            Try
                doSomething()
            Catch exc As ArgumentNullException 'Noncompliant
                Throw
            Catch exc As NotImplementedException
                Console.WriteLine("")
                Throw
            Catch exc As ArgumentException 'Noncompliant
                Throw
            Catch 'Noncompliant
                Throw
            End Try

            Try
                doSomething()
            Catch exc As ArgumentNullException 'Noncompliant
                Throw
            Catch exc As NotImplementedException
                Console.WriteLine("")
                Throw
            Catch exc As SystemException 'Noncompliant
                Throw
            Catch 'Noncompliant
                Throw
            End Try
        End Sub
    End Class

    ' Reproducer for https://github.com/SonarSource/sonar-dotnet/issues/8199
    Public Class Repro8163
        Public Sub SomeMethod()
            Throw New NotSupportedException()
        End Sub

        Public Function LogException(ex As Exception) As Boolean
            Return False
        End Function

        Public Sub CatchWithFilter()
            Try
                SomeMethod()
            Catch ex As Exception When LogException(ex)
                Throw
            End Try
        End Sub
    End Class
End Namespace

File: analyzers/tests/SonarAnalyzer.Test/TestCases/CatchRethrow.cs
using System;
using System.IO;

namespace Tests.TestCases
{
    class CatchRethrow
    {
        private void doSomething(  ) { throw new NotSupportedException()  ; }
        public void Test()
        {
            var someWronglyFormatted =      45     ;

            try
            {
                doSomething();
            }
            catch (Exception exc) //Noncompliant
            {
                throw;
            }

            try
            {
                doSomething();
            }
            catch (ArgumentException) { throw; } //Noncompliant {{Add logic to this catch clause or eliminate it and rethrow the exception automatically.}}
//          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

            try
            {
                doSomething();
            }
            catch (ArgumentException) //Noncompliant
            {
                throw;
            }
            catch (NotSupportedException) //Noncompliant
            {
                throw;
            }

            try
            {
                doSomething();
            }
            catch (ArgumentException)
            {
                throw;
            }
            catch
            {
                Console.WriteLine("");
                throw;
            }

            try
            {
                doSomething();
            }
            catch (ArgumentException) // Noncompliant
            {
                throw;
            }
            catch (NotSupportedException)
            {
                Console.WriteLine("");
                throw;
            }

            try
            {
                doSomething();
            }
            catch (ArgumentException) when (true)
            {
                throw;
            }
            catch (NotSupportedException)
            {
                Console.WriteLine("");
                throw;
            }

            try
            {
                doSomething();
            }
            catch (NotSupportedException) //Noncompliant
            {
                throw;
            }
            finally
            {

            }

            try
            {
                doSomething();
            }
            catch (ArgumentNullException) //Noncompliant
            {
                throw;
            }
            catch (NotImplementedException)
            {
                Console.WriteLine("");
                throw;
            }
            catch (ArgumentException) //Noncompliant
            {
                throw;
            }
            catch //Noncompliant
            {
                throw;
            }

            try
            {
                doSomething();
            }
            catch (ArgumentNullException)
            {
                throw;
            }
            catch (NotImplementedException)
            {
                Console.WriteLine("");
                throw;
            }
            catch (ArgumentException)
            {
                ;
                throw;
            }
            catch(SystemException) //Noncompliant
            {
                throw;
            }
            catch //Noncompliant
            {
                throw;
            }
        }
    }

    class CatchRethrowTemporaryContext
    {
        void OrdinaryUsing()
        {
            try
            {
                using (File.OpenRead("file"))
                {
                    throw new InvalidOperationException();
                }
            }
            catch // Noncompliant
            {
                throw;
            }
        }

        void LookAlikeMethods()
        {
            try
            {
                Impersonate();
                Run();
                Revert();
                throw new InvalidOperationException();
            }
            catch // Noncompliant
            {
                throw;
            }
        }

        void SpecificCatchBeforeRethrow()
        {
            try
            {
                throw new InvalidOperationException();
            }
            catch (ArgumentException)
            {
            }
            catch // Noncompliant
            {
                throw;
            }
        }

        void Impersonate() { }

        void Run() { }

        void Revert() { }
    }

    // Reproducer for https://github.com/SonarSource/sonar-dotnet/issues/8199
    public class Repro8199
    {
        public void SomeMethod() => throw new NotSupportedException();
        public bool LogException(Exception ex) => false;

        public void CatchWithFilter()
        {
            try
            {
                SomeMethod();
            }
            catch (Exception ex) when (LogException(ex))
            {
                throw;
            }
        }
    }
}
